instrument redis pipelined requests

krav opened the pull request:
<!-- If this pull request closes an issue, please mention the issue number below -->
Closes #BASE-137 

## 💸 TL;DR

We might want to drop the `command` label for these?

## 📜 Details
[Design Doc](<!-- insert Google Doc link here if applicable -->)

[Jira](<!-- insert Jira link if applicable -->)

<!-- Add additional details required for the PR: breaking changes, screenshots, external dependency changes -->

## 🧪 Testing Steps / Validation
<!-- add details on how this PR has been tested, include reproductions and screenshots where applicable -->

## ✅ Checks
<!-- Make sure your pr passes the CI checks and do check the following fields as needed - -->
- [ ] CI tests (if present) are passing
- [ ] Adheres to code style for repo
- [ ] Contributor License Agreement (CLA) completed if not a Reddit employee

krav committed:
instrument redis pipelined requests

krav committed:
add test

Xaelias commented on baseplate/clients/redis_cluster.py:
```diff
@@ -498,4 +498,38 @@ def execute_command(self, *args: Any, **kwargs: Any) -> Any:
+            commands = [cmd.args[0] for cmd in self.command_stack]
+
+            # Whether chained or pipelined it's one concurrent transaction
+            ACTIVE_REQUESTS.labels(**labels).inc()
```
https://github.com/reddit/baseplate.py/pull/731
you can do something like this for the active requests

krav replied:
Done

krav committed:
use track_inprogress

nsheaps commented on baseplate/clients/redis_cluster.py:
```diff
@@ -498,4 +498,37 @@ def execute_command(self, *args: Any, **kwargs: Any) -> Any:
+                success = "false"
+                raise
+            finally:
+                ACTIVE_REQUESTS.labels(**labels).dec()
```
the `with` block covers this case

krav replied:
as i was rewriting this i was thinking i'd leave something behind, and that i did

nsheaps commented on baseplate/clients/redis_cluster.py:
```diff
@@ -498,4 +498,37 @@ def execute_command(self, *args: Any, **kwargs: Any) -> Any:
+            commands = [cmd.args[0] for cmd in self.command_stack]
+
+            try:
+                with ACTIVE_REQUESTS.labels(**labels).track_inprogress():
```
[under the hood](https://sourcegraph.com/github.com/prometheus/client_python/-/blob/prometheus_client/context_managers.py?L37:7), this still uses `inc` and `dec`, can you please add a test for this (happy path and a case where `execute` returns an exception)

krav replied:
@Xaelias Any clue as how to do this? Looking at https://github.com/reddit/baseplate.py/pull/731/files#diff-a91d47420f289905a951c1be7b82eb613e3d3ca01e72871564b122a6a715774a the tests were just removed.

nsheaps replied:
I still don't see a test for this, I think I'd be satisfied by setup resetting the counter for calls and just checking that inc and dec were both called once and the resulting value is 0 (which you already have)

krav replied:
The old strategy to test doesn't seem to work with this. @Xaelias Did you spend any effort figuring out how to test this for #731? 

Happy to revert to using `inc` and `dec` as this PR originally had.

nsheaps commented on tests/unit/clients/redis_tests.py:
```diff
@@ -117,6 +127,15 @@ def test_execute_command(self, monitored_redis_connection, expected_labels):
+    def test_pipeline_instrumentation(self, monitored_redis_connection, expected_labels):
+        monitored_redis_connection.pipeline("test").set("hello", 42).set("goodbye", 23).execute()
+        labels = {**expected_labels, "redis_command": "SET", "redis_success": "true"}
+        assert REGISTRY.get_sample_value(f"{REQUESTS_TOTAL._name}_total", labels) == 2.0
```
I don't see a test for the `pipeline` part here

krav replied:
what do you mean? it's the same metric. the `.pipeline("test")` call instantiates the pipelined connection

nsheaps reviewed (changes requested):
Some stuff to address before this looks good to me

krav committed:
left over code was inevitable

KTAtkinson commented on tests/unit/clients/redis_tests.py:
```diff
@@ -117,6 +127,15 @@ def test_execute_command(self, monitored_redis_connection, expected_labels):
+        monitored_redis_connection.pipeline("test").set("hello", 42).set("goodbye", 23).execute()
+        labels = {**expected_labels, "redis_command": "SET", "redis_success": "true"}
+        assert REGISTRY.get_sample_value(f"{REQUESTS_TOTAL._name}_total", labels) == 2.0
+        assert (
```
Please provide a descriptive error message. This isn't the current style, but it helps if your test fails in the future.

krav replied:
Added

krav committed:
add assert description; active requests test

krav commented on baseplate/clients/redis.py:
```diff
@@ -269,7 +269,39 @@ def __init__(
+                success = "false"
+                raise
+            finally:
+                for command in commands:
```
Since no one's commented on it: This has a linear performance impact with the number of commands in the pipeline. It's probably not a noticeable impact, but I think it's unexpected from instrumentation. An option here is to deviate from the spec and drop the `command` label. The call would then just become a singular `.inc(len(self.command_stack))`.

nsheaps reviewed (changes requested):
Would like a test for the context manager handled track_inprogress call

krav committed:
revert use inprogress

krav committed:
test cluster instrumentation

krav commented:
> Would like a test for the context manager handled track_inprogress call

Changed it back to inc/dec with tests. Also added tests for the cluster client.

krav committed:
add assert error message to cluster tests

krav committed:
remove debug statement

nsheaps commented on baseplate/clients/redis.py:
```diff
@@ -269,7 +269,39 @@ def __init__(
+                        f"{PROM_LABELS_PREFIX}_success": success,
+                        f"{PROM_LABELS_PREFIX}_command": command,
+                    }
+                    REQUESTS_TOTAL.labels(**result_labels).inc()
```
Why are we doing this per command in the pipeline? Wouldn't the other commands be tracked separately from the pipeline itself via the `super` call?

nsheaps replied:
Separately I'm assuming that the active requests never track the individual commands within the pipeline since they're sent together?

krav replied:
> Why are we doing this per command in the pipeline? Wouldn't the other commands be tracked separately from the pipeline itself via the `super` call?

If you look at what that `super` call does it never hits any code path we've already instrumented. We assumed it did, which is why we didn't instrument this to begin with. What redis actually does is [pack the commands into one request](https://github.com/redis/redis-py/blob/f9f9d06c9951f8536bf9321dcebc96759eae03e0/redis/asyncio/client.py#L1261-L1266).

Hence its one "active request"/connection, but several commands being executed. See my other comment about including the individual names of commands executed, it's my preference to remove it but that's not compliant with the spec.

krav replied:
Actually, looking at the clustered Redis implementation (only), on retry it [can call `execute_command` on `RedisCluster`](https://github.com/redis/redis-py/blob/fd9fea6bc07bf0970a5a42c5ec1788272446910c/redis/cluster.py#L1998-L2026). As I understand it we want this to count in the metrics for every attempt, but potentially there are pitfalls.

nsheaps replied:
Cool, thanks for the explanation!

KTAtkinson replied:
Ideally, it doesn't make sense to associate the time the entire pipeline with each individual command in the pipeline. If folks want instrumentation per command they can provide a connection with instrumentation. Do you know how this work currently for Wavefront users?

nsheaps commented on tests/integration/redis_cluster_tests.py:
```diff
@@ -156,3 +157,12 @@ def test_pipeline(self):
+            "redis_command": "GET",
+            "redis_success": "true",
+        }
+        assert REGISTRY.get_sample_value("redis_client_requests_total", labels) == 5.0
```
Why test only the gets?

I feel like the test here should spell out what we expect to track, at minimum for clarity purposes:

total = 8
gets = 5
sets = 1
delete = 1
pipeline = 1

krav replied:
Aren't we then just testing the redis library? Every command follows the exact same code path in baseplate. It's passed through redis [pretty much always the same way](https://github.com/redis/redis-py/blob/031b2087b57713d29b98140fb1f0b9f7b4fc7a21/redis/commands/core.py#L1410-L1435), and back to our mocked connection object. 

Happy to add it, but would appreciate this kind of feedback up front.

nsheaps replied:
I can't provide feedback on code that hasn't been written yet :)

To be clear, what I'm calling out here is that you changed metrics, did a bunch of things to work through the pipeline code you wrote (which is great), but then only tested part of it. Tests should capture every aspect that we want to track, and in this case, it was a miss from the pipeline tracking. You got a test for tracking the gets within the pipeline, but not the pipeline tracking itself (which would fail because you never increment the total count for pipeline calls, nor are you tracking inc or dec for the active call) which would have been covered if writing assertions for the behavior that the test is performing (that was my original comment [here](https://github.com/reddit/baseplate.py/pull/732#discussion_r945893821) )

I'm not totally clear myself on what we want, which is partially why I'm asking for the tests that cover all the metrics we expect to be emitted from the test case here. Defining it as tests makes it much clearer than is defined elsewhere and easier to validate. Not defining it leads to ambiguity and potential for change without realizing that it's not fully covered by the unit test.

KTAtkinson replied:
TBH, this shouldn't be tacked onto an existing test, you should specifically have a test for metrics separate from any other test so that it's apparent what we're testing and what's not being tested.

krav replied:
Are all these things not all covered by the unit test? This should maybe not have made it into the PR. I added it for some assurance for myself early on and is now pretty redundant.

I don't understand your original comment, and you never followed it up. What does "the pipeline tracking itself" mean?

Looking at it, I think maybe there's a misunderstanding here that `pipeline()` is effectful? It's not, it just instantiates a `Pipeline(Redis)`. It's not until you call `execute()` that the pipeline start command `MULTI` is sent down the wire, so it's never counted against any of our metrics.

KTAtkinson commented on baseplate/clients/redis.py:
```diff
@@ -267,7 +267,39 @@ def __init__(
     def execute(self, **kwargs: Any) -> Any:
         with self.server_span.make_child(self.trace_name):
-            return super().execute(**kwargs)
+
```
nit: it looks like there is an additional blank line here

krav replied:
Weird this isn't caught by the formatter.

KTAtkinson commented on baseplate/clients/redis.py:
```diff
@@ -267,7 +267,39 @@ def __init__(
+                    "client_name", ""
+                ),
+                f"{PROM_LABELS_PREFIX}_database": self.connection_pool.connection_kwargs.get(
+                    "db", ""
```
Shouldn't this default to `0`?

krav replied:
This would then deviate from what we do in `MonitoredRedisConnection`.

KTAtkinson commented on baseplate/clients/redis_cluster.py:
```diff
@@ -497,4 +497,37 @@ def execute_command(self, *args: Any, **kwargs: Any) -> Any:
+                    "client_name", ""
+                ),
+                f"{PROM_LABELS_PREFIX}_database": self.connection_pool.connection_kwargs.get(
+                    "db", ""
```
Same as above, the default value here should be `0`.

krav replied:
Same here, we should change all of the defaults for redis in this case.

krav committed:
remove for loop

krav committed:
remove unused import
